useable version of pins_TEENSYLU.h

Bob-the-Kuhn opened the pull request:
1. corrected pin numbering
2. The endstop silkscreen doesn't agree with the schematic.  Modified the pin numbering so that activating the X stop (as shown on the silkscreen) tells the software that the X stop has been activated.
3. removed some defines that were causing the analog ins to be set to digital outputs.

See the new file for details.

Bob-the-Kuhn committed:
useable version of pins_TEENSYLU.h

1. corrected pin numbering
2. The endstop silkscreen doesn't agree with the schematic.  Modified
the pin numbering so that activating the X stop (as shown on the
silkscreen) tells the software that the X stop has been activated.
3. remove some defines that were causing the analog ins to be set to
digital outputs.

Bob-the-Kuhn commented:
Am I missing something?  The pins_PRINTRBOARD.h file also assumes that the "#define AT90USBxx_TEENSYPP_ASSIGNMENTS" line In FASTIO.h is commented out. 

Either there's a tool chain I haven't found or pins_PRINTRBOARD.h must be changed to the Teensy++2.0 pin mapping.

thinkyhead commented on Marlin/pins_TEENSYLU.h:
```diff
@@ -12,102 +12,314 @@
+ *      use the bootloader.
+ *   b) Download, if needed, a version of the Arduino IDE that is supported by the
+ *      "Teensyduino" library.  Supported versions are listed on the "Teensyduino"
+ *      download page.
```
Add the link here. http://www.pjrc.com/teensy/td_download.html

thinkyhead commented on Marlin/pins_TEENSYLU.h:
```diff
@@ -12,102 +12,314 @@
+
+
+  //
+  // Limit Switche definitions that match the schematic
```
"Switch"

thinkyhead commented:
This is really good work @Bob-the-Kuhn. It helps clarify the situation with pin mappings between "Marlin" and Teensy++2.0 (and I presume also 3.x?).

> The `pins_PRINTRBOARD.h` file also assumes that the
"`#define AT90USBxx_TEENSYPP_ASSIGNMENTS`" line
in `fastio.h` is commented out.

So it seems. All the Teensy-related pins files are:
```
pins_BRAINWAVE_PRO.h    : AT90USBxx_TEENSYPP_ASSIGNMENTS
pins_PRINTRBOARD.h      : Marlin (AT90USB) assignments
pins_PRINTRBOARD_REVF.h : Marlin (AT90USB) assignments
pins_TEENSY2.h          : AT90USBxx_TEENSYPP_ASSIGNMENTS
pins_TEENSYLU.h         : Marlin (AT90USB) assignments
```

thinkyhead commented:
Perhaps you can help me understand this table:
```
AT90USB  51 50 49 48 47 46 45 44 10 11 12 13 14 15 16 17 35 36 37 38 39 40 41 42 25 26 27 28 29 30 31 32 33 34 43 09 18 19 01 02 61 60 59 58 57 56 55 54
Port     A0 A1 A2 A3 A4 A5 A6 A7 B0 B1 B2 B3 B4 B5 B6 B7 C0 C1 C2 C3 C4 C5 C6 C7 D0 D1 D2 D3 D4 D5 D6 D7 E0 E1 E2 E3 E4 E5 E6 E7 F0 F1 F2 F3 F4 F5 F6 F7
Marlin   00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
Teensy   28 29 30 31 32 33 34 35 20 21 22 23 24 25 26 27 10 11 12 13 14 15 16 17 00 01 02 03 04 05 06 07 08 09(46*47)36 37 18 19 38 39 40 41 42 43 44 45
```
- The "Marlin ordering" is simply ordering the ports A-F, ascending 0-7, and counting those up from 0-47. We don't address these directly as pin numbers, but instead use `DIOn_PIN`. We might want to label that row "fastio" instead of "Marlin."
- The pin number that the processor recognizes for READ/WRITE is in the bottom row. So on the Teensy2 it's pin `28`.
- The "fastio number" 0 maps to the `A0` pin when `AT90USBxx_TEENSYPP_ASSIGNMENTS` is not defined (`#define DIO0_PIN PINA0`). That's _**physical**_ pin 51 on the AT90USB.
- The "fastio number" 0 maps to pin 0 (the `D0` pin) when `AT90USBxx_TEENSYPP_ASSIGNMENTS` is defined (`#define DIO0_PIN PIND0`).

Interestingly, the mappings when `AT90USBxx_TEENSYPP_ASSIGNMENTS` is enabled actually make the pins into direct untranslated pin numbers. That is, they negate the `fastio` translation, so `READ(0)` actually reads pin 0 on the processor (not whatever pin `A0` is). That is presumably port pin `D0` for the selected CPU.

~~So, in your table at the end of the `pins_TEENSYLU.h`, where you say that "`D0`" is the "AT90USB_ pin mapping", the actual pin should be "`D51`", right? (Because 0 is the mapping index, not the pin number.)~~

Likewise, while your Teensy pin mapping is technically correct, the reason the numbers given above work for `AT90USBxx_TEENSYPP_ASSIGNMENTS` is because they do map directly to the pin numbers.

So, perhaps it would be good to also include the mapping index as a column in that table.

thinkyhead commented:
```cpp

/**
 *
 *                      alternate  TEENSYPP       AT90USB…             name on
 *  names in this file  functions  pin numbering  pin numbering  port  silkscreen  schematic names
 *  X_STEP_PIN                     D28            D51            A0                X-STEP
 *  X_DIR_PIN                      D29            D50            A1                X-DIR
 *  Y_STEP_PIN                     D30            D49            A2                Y-STEP
 *  Y_DIR_PIN                      D31            D48            A3                Y-DIR
 *  Z_STEP_PIN                     D32            D47            A4                Z-STEP
 *  Z_DIR_PIN                      D33            D46            A5                Z-DIR
 *  E0_STEP_PIN                    D34            D45            A6                E-STEP
 *  E0_DIR_PIN                     D35            D44            A7                E-DIR
 *  SDSS                           D20            D10            B0                PB0-SS
 *                                 D21            D11            B1                SCLK
 *                                 D22            D12            B2                MOSI
 *                                 D23            D13            B3                MISO
 *  SPINDLE_SPEED_PIN   PWM2A      D24            D14            B4                EXP-PWM
 *  E_STOP_PIN                     D25            D15            B5    E-STOP      X-STOP
 *  X_STOP_PIN                     D26            D16            B6    X-STOP      Y-STOP
 *  Y_STOP_PIN                     D27            D17            B7    Y-STOP      Z-STOP
 *                                 D10            D35            C0                PC0
 *                                 D11            D36            C1                PC1
 *                                 D12            D37            C2                PC2
 *  E0_ENABLE_PIN                  D13            D38            C3                E-ENABLE
 *  HEATER_BED_PIN      PWM3C      D14            D39            C4                HOTBED
 *  HEATER_0_PIN        PWM3B      D15            D40            C5                HOTEND
 *  FAN_PIN             PWM3A      D16            D41            C6                PWM-FAN
 *  Z_ENABLE_PIN                   D17            D42            C7                Z-ENABLE
 *  CASE_LIGHT_PIN      PWM0B      D0             D25            D0                PD0-SCL
 *                      PWM2B      D1             D26            D1                PD1-SDA
 *  BTN_EN2                        D2             D27            D2                PD2-RX1
 *  BTN_EN1                        D3             D28            D3                PD3-TX1
 *                                 D4             D29            D4                PD4
 *                                 D5             D30            D5                PD5
 *                                 D6             D31            D6                PD6
 *                                 D7             D32            D7                PD7
 *                                 D8             D33            E0                PE0
 *                                 D9             D34            E1                PE1
 *                                 not assigned   D43            E2                ALE (PULL DOWN TO GROUND, JP32 TO +5)
 *                                 not assigned   D9             E3                to JP34
 *  Z_STOP_PIN                     D36            D18            E4    Z-STOP      E-STOP
 *                                 D37            D19            E5                PE5
 *  Y_ENABLE_PIN                   D18            D1             E6                Y-ENABLE
 *  X_ENABLE_PIN                   D19            D2             E7                X-ENABLE
 *  SDSS                A0         D38            D61            F0                A0
 *  SPINDLE_ENABLE_PIN  A1         D39            D60            F1                A1
 *  SPINDLE_DIR_PIN     A2         D40            D59            F2                A2
 *  BTN_ENC             A3         D41            D58            F3                A3
 *                      A4         D42            D57            F4                A4
 *                      A5         D43            D56            F5                A5
 *  TEMP_BED_PIN        A6         D44            D55            F6                B-THERM
 *  TEMP_0_PIN          A7         D45            D54            F7                E-THERM
 *
 */
```

Bob-the-Kuhn commented:
The top row are the physical pins on the IC package.

The Teensy row are the logical pin numbers used in the pins_xxx.h file when "AT90USBxx_TEENSYPP_ASSIGNMENTS" is enabled.

The Marlin row are the logical pin numbers used in the pins_xxx.h file when "AT90USBxx_TEENSYPP_ASSIGNMENTS" is disabled.

I don't see much value in including the physical package pins in the table.  I can certainly add them if you want.

I don't know of a method of compiling a useable image using the "Marlin" logical pin map.  I did spend some time modifying a couple of files that got me close.  There's some assembly language parts that I don't understand so I'm not confident that I have a robust "Marlin" solution.  I can spend more time on it if you see some value in it.

My vote would be to eliminate the "Marlin" pin mapping system and convert all the 90USB pins_xxxx.h files to use the Teensy pin mapping system.  I'd rather eliminate an unsupported system.  This would make things much simpler.

thinkyhead commented:
I see that in most cases in `fastio.h` we seem to be leaning in the direction of using pin mappings where the pin number _does_ correspond directly to the digital pin number. For example, when you look up fastio index 0 for ATmega2560 you get `PINE0`. And that _does_ function as digital pin `D0` on the ATmega2560.

The AT90USB seems to be the weird exception. It maps the indexes to the ports, in order, so there's no correspondence between the indexes and the digital pin numbers.

The pin numbers I referred to before for `AT90USB` are actually the physical pin numbers on the datasheet, the pins numbered from 1 - 64 in counter-clockwise ordering.

thinkyhead commented:
> I don't see much value in including the physical package pins in the table.

I agree. They just confuse things and should be removed altogether, or labeled better. We only care about how the pins are addressed in code.

thinkyhead commented:
> eliminate the "Marlin" pin mapping system and convert all the 90USB pins_xxxx.h files to use the Teensy pin mapping system.

I agree. The pin numbers we're using in most places correspond to the digital pin numbers. I think we should aim to get them all mapped that way.

Bob-the-Kuhn commented:
What should we do with FASTIO.h?  At the minimum we need to make "AT90USBxx_TEENSYPP_ASSIGNMENTS" default to enabled.

I think the best thing to do is just remove the "Marlin" system from FASTIO.

Bob-the-Kuhn committed:
simplify the changes

1. Removed the Rev A defines completely.  Keeping them around only
confuses things.

2. Changed the names for the two pin mapping systems to be more
descriptive and consistent.

3. Added a section to check the fuses before burning a new bootloader.

Bob-the-Kuhn commented:
I just pushed a new version of the file that incorporates the above comments.

Bob-the-Kuhn commented:
Sanity check - the two Printrboards both have the X_STOP_PIN assigned to port E3.  That port isn't assigned a logical pin in the Teensy library and is assigned logical pin 47 in FASTIO.

Will the X_STOP_PIN be functional if we use the Teensy settings in FASTIO?

All along I've assumed that the logical pin mapping in the library supersedes the mapping in FASTIO.  Does one always take precedence?

Blue-Marlin commented:
Whatever you change in fastio.h or the pin files - please check it with the relevant Arduino_pins.h of the relevant core/variant. As long as we have to use both, WRITE() and digitalWrite(), the mappings must be identical!

thinkyhead commented:
> As long as we have to use both, WRITE() and digitalWrite(), the mappings must be identical!

Is that the reason why we use AT90USB mapping in some cases and Teensy (direct) mapping in other cases? To comply with the `WRITE` and `digitalWrite` parameters? I'm surprised that they would differ. Where do they take different numbers?

thinkyhead commented:
`Arduino_pins.h` rules them all.

Bob-the-Kuhn commented:
I think I've figured out what's going on with Teensylu & the Printrboards.

They don't use standard Marlin.  They're actually using both the Teensyduino and the FASTIO pin mapping systems.

They depend on a pin only being accessed by the FASTIO `READ` & `WRITE` macros or by the `digitalRead`/`digitalWrite` functions but not both.

That would explain the strange comments in the pins_xxx.h files.  They're trying to tell you which system that pin is using.

I've asked this question in a couple of forums & hope to get confirmation in a day or two.

thinkyhead commented:
So, maybe a translation layer is needed to convert from one to the other, then we can go back to a pins file with only one set of assignments needed. By digital pin numbers would still be my vote.

Bob-the-Kuhn commented:
I found Arduino extensions for the "standard Marlin" pin map and for the 48 pin version of the Teensy pin map.  

There was a link in the pins_BRAINWAVE.h file to https://github.com/unrepentantgeek/brainwave-arduino.  That link had an Arduino 1.0.5 extension and an unreleased Arduino 1.5 extension.

Looks like we have several options.  None of them are perfect.  One of the big issues is getting the compiler flags/options set so that the latest Marlin code will compile without errors.  So far they've mostly been requests for syntax changes or to explicitly do a type conversion.

I'm going to put together a list of the options along with the pros & cons.  I hope to have it tomorrow.

Bob-the-Kuhn commented:
I have a version of the Printrboard Arduino extension 99% working.  The only mods I had to make were additions to the boards.txt file.  I think this is a good solution for the boards that do not use the E3 port.

The E3 boards (5DPRINT, PRINTRBOARD & PRINTRBOARD_REVF) don't have an elegant solution.  The one I like best is to use a moderately modified version of the Dwengino extension (I found it today).  It needs more testing before I call it done. The upside is it supports all 48 pins. The down side is where to save it so that the users can easily download it.

If you don't like the Dwengino option then we're stuck with 46 pin options.  

On the 5DPRINT board the E3 port is used for SD card detection.  We can live without that.

The PRINTRBOARDs use E3 for the X_STOP_PIN.  The options I see are:
- use the current system but do a better job of describing it and the risks that result
- use only the Teensy pin assignments which means the user would have to do some custom cabling to use the X_STOP
- offer all three options (Dwengino, current system & Teensy pin map only)

The reason why I don't want to use the Teensyduino extension is it forces the user to use the USBtinyISP to upload new code.  The Printrboard extension has a nice bootloader system.  The USBtinyISP is only needed to get the bootloader on to the board.  After that it works just like the Arduino system (just click the compile & upload button).

Bob-the-Kuhn commented:
I've got the modified Dwengino running to my satisfaction.  It supports both the Tweensy 48 pin map and the "standard" pin map.  This system can be used on all our AT90USB controllers.

I'm sure there's lots of areas that can be improved in the system I have currently.  I know that the compiler options can be better because there's lots of warnings and I had to make mods to the pinsDebug.h file to get it to compile.




The biggest hurdle is where to put it so it's easily accessible.

Some finer points that I need help with:
- There are USB items like PID, VID & manufacturer.  I'll need help sorting them out.  I don't know what's reasonable for them.  If we can't just hardcode in some values then I'll need help with the compiler options in the platforms.txt file.  The link between it and the compiler is broken & I haven't the foggiest.
- Is there a way of querying the board name (as shown in the Arduino IDE)?  I'm wondering if there is a way to protect the user from selecting an invalid combination of pin map and  AT90USBxx_TEENSYPP_ASSIGNMENTS setting.  Probably best to just warn them in the pinsxxx.h file.

I found out what happens when an  invalid combination is selected.  Not much works.  I was pulling my hair out for most of the day because I had forgotten to set AT90USBxx_TEENSYPP_ASSIGNMENTS  correctly.

I've attached the modified Dwengino system.
[Marlin_AT90USB.zip](https://github.com/MarlinFirmware/Marlin/files/656575/Marlin_AT90USB.zip)

thinkyhead commented:
> I was pulling my hair out for most of the day

The `AT90USBxx_TEENSYPP_ASSIGNMENTS` flag can't be overridden early, since `fastio.h` is one of the first included files. So the best place to check for the correctly-chosen mapping is in the relevant `pins_MYBOARD.h` files.

Bob-the-Kuhn commented:
My first attempt was declaring a flag in the pins_arduino.h file in the variants directory and then querying it in the pins_xxx.h file.  Turns out the flag passes the #ifdef test but the value is always zero until later in the compilation process.  Drove me nuts trying to figure out why my if.. elif chain was not working consistently.

I was able to create a system that works by creating the flag in the platforms.txt file and then giving it a value in the boards.txt file in the same section as the variant is specified.

I'm scratching my head as to if protecting the user is worth the effort.  The alternative is to have the user select (via a define) which option they want and hope they don't shoot themselves in the foot too often.

I was able to find an Arduino IDE extension that supported the  Sanguinololu type boards with Arduino 1.6.11 and later.  I was able to add the Sanguinololu pin map variant to it and compile RCBugFix with it.  The downside is the extension author/maintainer isn't interested in supporting the Sanguinololu pin map.

So now I have an orphan to support the Teensylu family and another orphan to support the Sanguinololu family.

3/4 of the boards couldn't compile RCBugFix with the tools that were available 2 weeks ago.  That implies there's almost no interest in using these boards with the current generation of Marlin.  Hmmm - could be chicken & egg situation here.  My impression is that there just isn't that many people using these old boards.

My suggestion is to update the pins_xxx.h files with pointers to the orphans (and some instructions) and then declare end of support for them.

What do you want done?

Is there a logical home within the Marlin repository structure for the two orphans?  I don't mind putting them in my GitHub area as long as we make it clear to the users that there will be minimal, if any, support.

thinkyhead commented:
> Drove me nuts trying to figure out why my if.. elif chain was not working consistently.

Yep, the preprocessor is an interesting beast. (See how we use `thermistornames.h`, for example.)

> I was able to create a system that works by creating the flag in the platforms.txt file and then giving it a value in the boards.txt file in the same section as the variant is specified.

Obviously not a user-friendly option. From our side, we just need to throw an error if something is missing or mis-defined. Directing users to edit `fastio.h` is still nicer than requiring edit of `boards.txt` / `platforms.txt`.

> My suggestion is to update the pins_xxx.h files with pointers to the orphans (and some instructions) and then declare end of support for them.

I'd like to see what that entails. Can we see it as a PR — or adding onto this one?

If we can support the boards at an equal level that would be best. In the case of the original concept —having two pin mappings in the same `pins_MYBOARD.h` file— it might be just as well to split it up into two separate pins files, each with its own ID number.

Because of the extra work required —obtaining boards files and setting up Arduino IDE— we should post instructions for any special builds (Teensy, Sanguinololu, etc.) right onto the website, and just include a basic explanation and links in the pins files.

Bob-the-Kuhn commented:
The editing of the boards.txt and platform.txt files would only be done by the person(s) setting up the extension/variant.  Assuming the people setting up the extension/variant talk to the people maintaining the pins_MYBOARD.h  files, then all the user needs to do is to select the variant they want from the Arduino IDE.  The pins file will query the variant flag and the AT90USBxx_TEENSYPP_ASSIGNMENTS setting and then select the appropriate set of defines.

Probably better in the long run just to have instructions in the pins file saying something like:
  If you want a), b) & c) to happen then:
       1. uncomment the "#define option A" line in this file
       2. uncomment AT90USBxx_TEENSYPP_ASSIGNMENTS flag in FASTIO.h
      3. select board XYZ in the Arduino IDE
  If you want a) to happen  but not b) and c) then:
       1. uncomment the "#define option B" line in this file
       2. uncomment AT90USBxx_TEENSYPP_ASSIGNMENTS flag in FASTIO.h
      3. select board ABC in the Arduino IDE

Bob-the-Kuhn commented:
I've rolled this in with another PR
